_source/_posts/2022-01-04-angular-auth0-quickly.md

```diff
@@ -0,0 +1,177 @@
+google.charts.load('current', {packages: ['corechart', 'bar']});
+google.charts.setOnLoadCallback(drawChart);
+function drawChart() {
+  var data = google.visualization.arrayToDataTable([
+    ['Framework', 'Stack Overflow tags', { role: 'style' }],
+    ['Angular', 266938, 'red'],
+    ['React', 348946, '#61dafb'],
+    ['Vue', 87085, '#4fc08d']
+  ]);
+  var options = {
+    title: 'Stack Overflow Tagged Questions',
+    chartArea: {width: '50%'},
+    hAxis: {
+      title: 'Tags',
+      minValue: 0
+    },
+    vAxis: {
+      title: 'Web Framework'
+    }
+  };
+  var chart = new google.visualization.BarChart(document.getElementById('stack-overflow-tags'));
+  chart.draw(data, options);
+}
+</script>
+
+I kid, I kid. 😅
+
+I don't think Stack Overflow tags indicate that developers have a hard time with the framework, I think they indicate a thriving community. The more questions, the more people are using it. 
+
+[hotframeworks.com](https://hotframeworks.com/) lists a score that combines GitHub stars with Stack Overflow tags and says the top six web frameworks are as follows (at the time of this writing):
+
+1. React
+2. ASP.NET MVC
+3. Angular
+4. Ruby on Rails
+5. AngularJS
+6. Vue.js
+
+Angular is still very popular among developers. If you're developing an Angular app today, you probably need a way to authenticate your users. That's where OpenID Connect (OIDC) can help you. OIDC is a layer on top of OAuth 2.0 that provides identity. 
+
+In this tutorial, I'll how you to add OIDC authentication with Auth0 to a new Angular app in just a few steps. 
+
+**Prerequisites:**
+
+* A favorite text editor or IDE. I recommend [IntelliJ IDEA](https://www.jetbrains.com/idea/), but I know many JavaScript developers prefer [Visual Studio Code](https://code.visualstudio.com/).
+* [Node.js](http://nodejs.org/) v14+ and npm installed.
+
+{% include toc.md %}
+
+## Create an Angular app
+
+First, you'll need to create a new Angular app with routing enabled.
+
+```shell
+npx @angular/cli@13 new auth0-demo --routing
+```
+
+Select your favorite stylesheet format. The selection you make doesn't matter for this example.
+
+## Add Auth with OpenID Connect
+
+To add authentication with Auth0, you'll first need a [free Auth0 account](https://auth0.com/signup). Install the [Auth0 CLI](https://github.com/auth0/auth0-cli#installation) and run `auth0 login` to register your account. Then, run `auth0 apps create`. Specify a name and description of your choosing. Select **Single Page Web Application** and use `http://localhost:4200/home` for the Callback URL. Specify `http://localhost:4200` for the rest of the URLs.
+
+{% img blog/angular-auth0/auth0-apps-create.png alt:"auth0 apps create" width:"800" %}{: .center-image }
+
+Once you have a new Angular app and Auth0 OIDC set up, you can use [OktaDev Schematics](https://github.com/oktadev/schematics#angular--auth0) to add OAuth 2.0 and OIDC support to your Angular app.
+
+```shell
+ng add @oktadev/schematics --auth0
```

Comment:
I had fairly severe issues with the Angular CLI here and the default app scaffold. 

`npx @angular/cli@13 new auth0-demo --routing` shot off an error about an incompatibility between jasmine packages. You can find a github issue [here ](https://github.com/angular/angular-cli/issues/20719) that's marked as resolved, but it really isn't.

In addition the default app set up, bizarrely, uses v11 cli and core for angular. It also has 64 high vulnerabilities out of the gate.

In order to get the oktadev schematic to install correctly I had to run:

1. npm update
2. npm audit fix --force
2. ng update @angular/cli@12 --force
3. ng update@angular/cli@13

Reply:
@Ryanman What version of Node.js and npm are you using? Also, which operating system? As you can see from the screencast I created, I did not experience any of these issues.

Reply:
I just tried `npx @angular/cli@13 new auth0-demo --routing` and `ng add @oktadev/schematics --auth0` again and didn't experience any issues. I'm using Node.js 14 and npm 8.

<img width="1457" alt="npx @angular:cli" src="https://user-images.githubusercontent.com/17892/148810951-f96ea83b-b52a-41df-ad68-1e954ca7c8ca.png">

<img width="1457" alt="ng add @oktadev:schematics" src="https://user-images.githubusercontent.com/17892/148810997-e141073c-3b80-4ce9-be96-a9fcca030a4d.png">

Reply:
This is crazy! I've chosen the SCSS processor, and yes to TS. However I still get the error with vanilla JS and CSS.

![image](https://user-images.githubusercontent.com/1376624/148846216-b71f696e-ead7-4983-93cd-f060343d68be.png)

I am on windows, but I'm having a hard time thinking that's the issue. Perhaps we mention it as a potential error to account for? I'll try on a different machine soon.

![image](https://user-images.githubusercontent.com/1376624/148846529-dc95ef24-f7db-4c59-8e80-4331af68c31c.png)

Reply:
Can you try with Node 14 and see if it still happens? I'll try with Node 16 too.

Reply:
Are you running `npx @angular/cli@13 new auth0-demo --routing` or `ng new`? I tried it with Node.js 16 and didn't have any issues. You might try `npm uninstall -g @angular/cli`?

I have a Windows 10 machine in Parallels I'll try...

Reply:
I tried running `npx @angular/cli@13 new auth0-demo --routing` on my Parallels instance of Windows 10 and it works fine with Node.js 16. 

<img width="1464" alt="Screen Shot 2022-01-10 at 22 35 54" src="https://user-images.githubusercontent.com/17892/148887428-c305c240-96d2-4ee0-89d6-f30db413726c.png">

However, when I try to run `ng add`, I do get an error:

```
C:\Users\mraible>cd auth0-demo

C:\Users\mraible\auth0-demo>ng add @oktadev/schematics --auth0
'ng' is not recognized as an internal or external command,
operable program or batch file.
```

If I use `npx @angular/cli@13 add @oktadev/schematics --auth0`, it works. Maybe it'd be best to just recommend `npm install -g @angular/cli@13` and not use `npx`?

Reply:
I tested this on a different machine and had 0 issues. I then updated my global install of the angular CLI and encountered no problems. 

I apologize for the heartburn! We should be good.

Reply:
@mraible if you don't have Angular CLI globally installed, you'll have to use `npx` for all `ng` commands, as you noted for adding the schematic (same with `npx @angular/cli generate component login` for example).

So I recommend still having a version of Angular CLI installed. But by using `npx` to create the app, you can specify the version of the Angular app and you'll use the project version of the CLI tool for subsequent commands in the project dir. Angular used to output a warning about that in the `ng --version` command, but I don't see it anymore now. However, you can see the difference in version output. In the screenshots below I have Angular v10 globally installed and created the app with Angular v13.

![v10](https://user-images.githubusercontent.com/7586237/148955130-dc9647f2-90d3-4ac2-be90-aa773d866ef2.jpg)
![v13](https://user-images.githubusercontent.com/7586237/148955128-90bf3fe9-555d-4eba-aeb0-51c2533cc1c7.jpg)

IMHO - I think it's a good idea to use `npx`. This is so people don't have to keep changing their Angular CLI version to follow a tutorial. On my work machine previously, I wanted to keep my Angular CLI/Node/npm version to match my build server and want to streamline work needs, not tutorial requirements.

Reply:
I haven't really started using `npx` until the last year. Since it causes confusion in this post, I decided to remove it.